theano/tensor/signal/pool.py

```diff
@@ -310,23 +366,25 @@ def perform(self, node, inp, out):
         elif self.mode != 'max':
             func = numpy.average
 
-        for n in xrange(x.shape[0]):
-            for k in xrange(x.shape[1]):
-                for r in xrange(pr):
-                    row_st = r * st0
-                    row_end = builtins.min(row_st + ds0, img_rows)
-                    if not inc_pad:
-                        row_st = builtins.max(row_st, self.padding[0])
-                        row_end = builtins.min(row_end, x.shape[-2] + pad_h)
-                    for c in xrange(pc):
-                        col_st = c * st1
-                        col_end = builtins.min(col_st + ds1, img_cols)
-                        if not inc_pad:
-                            col_st = builtins.max(col_st, self.padding[1])
-                            col_end = builtins.min(col_end,
-                                                   x.shape[-1] + pad_w)
-                        zz[n, k, r, c] = func(y[
-                            n, k, row_st:row_end, col_st:col_end])
+        # precompute the region boundaries for each dimension
```

Comment:
The reason I did that was mainly because I made the unoptimized version really, really slow, which would mess up the tests. Computing the list of `slice` objects for each region takes a lot of time, especially in 3D.